### PR TITLE
APIGOV-23845 - use atomic uint to lock/unlock publishing/validation

### PR DESCRIPTION
**Pull request by jcollins-axway**

### CODE REVIEW

**Review by sbolosan — APPROVED** [review]
🤙🏾

File: pkg/agent/instancevalidator.go [hunk 1]
```diff
@@ -51,66 +72,95 @@ func (j *instanceValidator) validateAPIOnDataplane() {
...
 			continue
 		}
+		logger = logger.WithField("name", instance.Name)
 
 		externalAPIID, _ := util.GetAgentDetailsValue(instance, defs.AttrExternalAPIID)
 		if externalAPIID == "" {
+			logger.Trace("could not get instance external id")
 			continue // skip service instances without external api id
+		} else if err != nil {
+			logger.WithError(err).Trace("could not get instance external id")
```
[the remark this answers is not in the document]
Comment (jcollins-axway) [reply pkg/agent/instancevalidator.go:100]: I actually don't think this conditional will ever be hit, if externalAPIID is "" and there was or wasn't an error then the if block would catch it. There will never be a time err is not nil and externalAPIID is set.

File: pkg/agent/instancevalidator.go [hunk 2]
```diff
@@ -51,66 +72,95 @@ func (j *instanceValidator) validateAPIOnDataplane() {
...
+			logger = logger.WithField("externalAPIStage", externalAPIStage)
+		}
 		externalPrimaryKey, _ := util.GetAgentDetailsValue(instance, defs.AttrExternalAPIPrimaryKey)
+		if externalPrimaryKey != "" {
+			logger = logger.WithField("externalPrimaryKey", externalPrimaryKey)
+		}
 		// Check if the consumer instance was published by agent, i.e. following attributes are set
 		// - externalAPIID should not be empty
 		// - externalAPIStage could be empty for dataplanes that do not support it
+		logger.Trace("validating API Instance on dataplane")
```
[the remark this answers is not in the document]
Comment (jcollins-axway) [reply pkg/agent/instancevalidator.go:114]: this is specifically for a single instance, 77 is for all instanaces

[the remark this answers is not in the document]
Comment (sbolosan) [reply pkg/agent/instancevalidator.go:114]: 👍🏾

File: pkg/agent/instancevalidator.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (sbolosan) [reply pkg/agent/instancevalidator.go:100]: Solid

**Review by sbolosan — APPROVED** [review]

**Review by vivekschauhan — APPROVED** [review]

**Pull request merged**